File: akka-actor-typed-tests/src/test/scala/docs/akka/typed/fromclassic/TypedSample.scala
/*
 * Copyright (C) 2019-2023 Lightbend Inc. <https://www.lightbend.com>
 */

package docs.akka.typed.fromclassic

// #hello-world-actor
import akka.actor.typed.ActorRef
import akka.actor.typed.Behavior
import akka.actor.typed.scaladsl.AbstractBehavior
import akka.actor.typed.scaladsl.ActorContext
import akka.actor.typed.scaladsl.Behaviors

// #hello-world-actor

object TypedSample {

  // #hello-world-actor
  object HelloWorld {
    final case class Greet(whom: String, replyTo: ActorRef[Greeted])
    final case class Greeted(whom: String, from: ActorRef[Greet])

    def apply(): Behavior[HelloWorld.Greet] =
      Behaviors.setup(context => new HelloWorld(context))
  }

  class HelloWorld(context: ActorContext[HelloWorld.Greet]) extends AbstractBehavior[HelloWorld.Greet](context) {
    import HelloWorld._

    override def onMessage(message: Greet): Behavior[Greet] = {
      context.log.info("Hello {}!", message.whom)
      message.replyTo ! Greeted(message.whom, context.self)
      this
    }
  }
  // #hello-world-actor

  // #children
  object Parent {
    sealed trait Command
    case class DelegateToChild(name: String, message: Child.Command) extends Command
    private case class ChildTerminated(name: String) extends Command

    def apply(): Behavior[Command] = {
      def updated(children: Map[String, ActorRef[Child.Command]]): Behavior[Command] = {
        Behaviors.receive { (context, command) =>
          command match {
            case DelegateToChild(name, childCommand) =>
              children.get(name) match {
                case Some(ref) =>
                  ref ! childCommand
                  Behaviors.same
                case None =>
                  val ref = context.spawn(Child(), name)
                  context.watchWith(ref, ChildTerminated(name))
                  ref ! childCommand
                  updated(children + (name -> ref))
              }

            case ChildTerminated(name) =>
              updated(children - name)
          }
        }
      }

      updated(Map.empty)
    }
  }
  // #children

  object Child {
    sealed trait Command

    def apply(): Behavior[Command] = Behaviors.empty
  }

}

File: akka-actor-typed-tests/src/test/java/jdocs/akka/typed/fromclassic/TypedSample.java
/*
 * Copyright (C) 2019-2022 Lightbend Inc. <https://www.lightbend.com>
 */

package jdocs.akka.typed.fromclassic;

// #hello-world-actor
import akka.actor.typed.ActorRef;
import akka.actor.typed.Behavior;
import akka.actor.typed.javadsl.AbstractBehavior;
import akka.actor.typed.javadsl.ActorContext;
import akka.actor.typed.javadsl.Behaviors;
import akka.actor.typed.javadsl.Receive;

import java.util.HashMap;
import java.util.Map;

// #hello-world-actor

interface TypedSample {

  // #hello-world-actor
  public class HelloWorld extends AbstractBehavior<HelloWorld.Greet> {

    public static final class Greet {
      public final String whom;
      public final ActorRef<Greeted> replyTo;

      public Greet(String whom, ActorRef<Greeted> replyTo) {
        this.whom = whom;
        this.replyTo = replyTo;
      }
    }

    public static final class Greeted {
      public final String whom;
      public final ActorRef<Greet> from;

      public Greeted(String whom, ActorRef<Greet> from) {
        this.whom = whom;
        this.from = from;
      }
    }

    public static Behavior<Greet> create() {
      return Behaviors.setup(HelloWorld::new);
    }

    private HelloWorld(ActorContext<Greet> context) {
      super(context);
    }

    @Override
    public Receive<Greet> createReceive() {
      return newReceiveBuilder().onMessage(Greet.class, this::onGreet).build();
    }

    private Behavior<Greet> onGreet(Greet command) {
      getContext().getLog().info("Hello {}!", command.whom);
      command.replyTo.tell(new Greeted(command.whom, getContext().getSelf()));
      return this;
    }
  }
  // #hello-world-actor

  // #children
  public class Parent extends AbstractBehavior<Parent.Command> {

    public interface Command {}

    public static class DelegateToChild implements Command {
      public final String name;
      public final Child.Command message;

      public DelegateToChild(String name, Child.Command message) {
        this.name = name;
        this.message = message;
      }
    }

    private static class ChildTerminated implements Command {
      final String name;

      ChildTerminated(String name) {
        this.name = name;
      }
    }

    public static Behavior<Command> create() {
      return Behaviors.setup(Parent::new);
    }

    private Map<String, ActorRef<Child.Command>> children = new HashMap<>();

    private Parent(ActorContext<Command> context) {
      super(context);
    }

    @Override
    public Receive<Command> createReceive() {
      return newReceiveBuilder()
          .onMessage(DelegateToChild.class, this::onDelegateToChild)
          .onMessage(ChildTerminated.class, this::onChildTerminated)
          .build();
    }

    private Behavior<Command> onDelegateToChild(DelegateToChild command) {
      ActorRef<Child.Command> ref = children.get(command.name);
      if (ref == null) {
        ref = getContext().spawn(Child.create(), command.name);
        getContext().watchWith(ref, new ChildTerminated(command.name));
        children.put(command.name, ref);
      }
      ref.tell(command.message);
      return this;
    }

    private Behavior<Command> onChildTerminated(ChildTerminated command) {
      children.remove(command.name);
      return this;
    }
  }
  // #children

  public class Child {
    public interface Command {}

    public static Behavior<Command> create() {
      return Behaviors.empty();
    }
  }
}
